v03_pipeline/lib/tasks/reference_data/updated_reference_dataset_collection_task.py

```diff
@@ -0,0 +1,206 @@
+from v03_pipeline.lib.tasks.base.base_update_task import BaseUpdateTask
+from v03_pipeline.lib.tasks.files import GCSorLocalTarget
+
+# NOTE: this is intended to replace the code in v03_pipeline/bin/write_combined_reference_ht.py, v03_pipeline/bin/write_combined_interval_reference_ht.py, and combine.py
+
+class UpdatedReferenceDatasetCollectionTask(BaseUpdateTask):
+    reference_dataset_collection = luigi.EnumParameter(enum=ReferenceDatasetCollection)
+    dataset = luigi.OptionalStrParameter()
+
+    @cached_property
+    def _destination_path(self) -> str:
+        return valid_reference_dataset_collection_path(
+            self.reference_genome,
+            self.dataset_type,
+            self.reference_dataset_collection,
+        )
+
+    def output(self) -> luigi.Target:
+        return GCSorLocalTarget(self._destination_path)
+
+    def initialize_table(self) -> hl.Table:
+        if self.dataset is None:
+            key_type = self.reference_dataset_collection.table_key_type(self.reference_genome)
+            return hl.Table.parallelize(
+                [],
+                key_type,
+                key=key_type.fields,
+                globals=hl.Struct(
+                    paths=hl.Struct(),
+                    versions=hl.Struct(),
+                    enums=hl.Struct(),
+                ),
+            )
+
+        if not hl.hadoop_exists(os.path.join(self._destination_path, '_SUCCESS')):
+            message = (
+                f'Cannot find existing valid ht at ({self._destination_path}). '
+                f'Consider omitting the --dataset flag to re-initialize the ht.'
+            )
+            raise Exception(message) # TODO: new exception type?
+
+        else:
+            return hl.read_table(self._destination_path)
+
+    def update_table(self, ht: hl.Table) -> hl.Table:
+        joined_ht = ht
+
+        if self.dataset is None:
+            for dataset in self.reference_dataset_collection.datasets(self.dataset_type):
+                dataset_ht = self._get_dataset_ht(dataset)
+                joined_ht = joined_ht.join(dataset_ht, 'outer')
+                joined_ht = self._annotate_dataset_globals(joined_ht, dataset, dataset_ht)
+
+        elif self.dataset is not None:
+
+            # TODO check globals for out of sync dataset references?
+
+            dataset_ht = self._get_dataset_ht(self.dataset)
+            joined_ht = joined_ht.drop(self.dataset)
+            joined_ht = joined_ht.join(dataset_ht, 'outer')
+            joined_ht = joined_ht.filter(
+                hl.any(
+                    [
+                        ~hl.is_missing(joined_ht[dataset])
+                        for dataset in self.reference_dataset_collection.datasets(self.dataset_type)
+                    ],
+                ),
+            )
+            joined_ht = self._annotate_dataset_globals(joined_ht, self.dataset, dataset_ht)
+
+        return joined_ht
+
+    def _get_dataset_ht(self, dataset: str) -> hl.Table:
```

Comment:
This is getting more opinion-ey, but I'd opt for leaving the rest of these functions in the `lib/reference_data` folder, but probably in something not named `combine.py`.  They're more general than this task in particular, and I have a feeling one or two will be be used in other tasks.  We can also just leave the existing unit tests that we have working.  

I'm a fan of the rename from `get_ht` to `get_dataset_ht` though, that's for sure a win.

Reply:
This is something considered and I agree here - to keep these operations separate from the luigi task as they are now. I'll add the single function to create/update the joined table in that file. I've settled on renaming the file `dataset_table_operations.py` for now but am open to a different name